text/0779-first-class-component-templates.md

```diff
@@ -0,0 +1,1447 @@
+
+    ```js
+    function conditionalComponent(predicate) {
+      if (predicate) {
+        return <template><p>Cool</p></template>
+      } else {
+        return <template><p>Lame</p></template>
+      }
+    }
+    ```
+
+    However, doing so will have unexpected costs at runtime. It’s worth remembering what the resulting output is:
+
+    ```js
+    function conditionalComponent(predicate) {
+      if (predicate) {
+        return setComponentTemplate(
+          precompileTemplate(`<p>Cool</p>`),
+          templateOnly()
+        );
+      } else {
+        return setComponentTemplate(
+          precompileTemplate(`<p>Lame</p>`),
+          templateOnly()
+        );
+      }
+    }
+    ```
+
+    The problem here is that this requires re-running both the creation of the template-only empty backing instance which has `null` for `this`, and the association between the two. (Normally the template precompilation also happens during the buidl, which eliminates some but not all of the apparent cost here. Re-doing the template precompilation will *work*, but it is expensive and slow, so we do it during build.)
+
+    In this scenario, users can accomplish the same thing by manually hoisting the component definitions to module scope:
+
+    ```js
+    const Cool = <template><p>Cool</p></template>;
+    const Lame = <template><p>Lame</p></template>;
+
+    function conditionalComponent(predicate) {
+      if (predicate) {
+        return Cool;
+      } else {
+        return Lame;
+      }
+    }
+    ```
+
+    (Notice that this is no different than any other concern around other costly operations needlessly happening in a function body repeatedly.)
+
+    See further discussion below under [**How We Teach This**](#how-we-teach-this).
+
+
+#### Standalone
+
+The compiled output for a top-level `<template>` tag *not* bound to a name is a default export, but otherwise identical to a `<template>` declaration bound to a tag. Thus, given this input:
+
+```js
+<template>
+  <p>Hello, {{@name}}!</p>
+</template>
+```
+
+The compiled output is:
+
+```js
+import { precompileTemplate } from '@ember/template-compilation';
+import { setComponentTemplate } from '@ember/component';
+import templateOnlyComponent from '@ember/component/template-only';
+
+export default setComponentTemplate(
+  precompileTemplate(`
+  <p>Hello, {{@name}}!</p>
+  `
+  ),
+  templateOnlyComponent()
+);
+```
+
+Scoped values are included the same as with named definitions.
+
+Since the compiled output is a default export, it is a static error to have multiple top-level (i.e. not bound to a name) `<template>`s in a file—because it is a static error to have multiple `export default` statements in a JavaScript file. (We should provide linting tooling to error on this case, rather than letting it fail at runtime.)
```

Comment:
out of curiosity what's the reason on making these default exports? wouldn't making them non-default exports avoid this issue and the need for a lint rule? I'm guessing there's some good reasons for this though, so overall just curious!

Reply:
It means that a simple template-only component is basically “just HTML”. There’s no need for `export` for the simplest cases (which is *many* of them). As a number of the code samples throughout demonstrate, the result is a design where you can legitimately start with plain HTML (just wrapped in `<template>`) and progressively add functionality via JS inports without needing to change the original declaration and without needing to add JS markup to the template itself. This is better for teaching it, and it also looks and feels quite nice—which is definitely lowest on my priority list for evaluating this, but all other things being equal…

Reply:
I re-arranged the sections and called this out explicitly so that it’s clearer to folks who haven’t been deep in these design weeds for as long as I have. Thanks for asking about it!